Support ADD COLUMN, ALTER PRIMARY KEY, DROP COLUMN in one stmt

Xiang-Gu opened the pull request:
This PR laid out foundations to support DDL stmt that requires constructing more than one primary index. In particular, this work is centered around supporting ADD COLUMN(s), ALTER PRIMARY KEY, DROP COLUMN(s) in one ALTER TABLE statement, which would require constructing at least one, and at most three primary indexes to achieve a correct online schema change.

The general idea to achieve this is to construct a sequence of four primary indexes, each sourced from the previous one, `old` <-- `inter1` <-- `inter2` <-- `final`, where
 - `old` is the existing, public primary indexes;
 - `inter1` is the one that includes all the to-be-added and to-be-dropped columns;
 - `inter2` has the same columns as `inter1` but has its key columns altered (to handle the ALTER PRIMARY KEY command)
 - `final` is the final primary index with the correct set of columns and correct key after completing the schema change.

For each ALTER TABLE stmt, if we determined that new primary index is needed, we initialize this sequence of primary indexes by cloning from `old` (this action is also known as "inflating a chain of primary indexes"). Once we have all four primary indexes, we process each ADD COLUMN, ALTER PRIMARY KEY, and DROP COLUMN command by modifying some of the primary indexes:
 - ADD COLUMN: add the column as a stored column to `inter1`, `inter2`, and `final`;
 - ALTER PRIMARY KEY: alter the key/stored columns in `inter2` and `final`;
 - DROP COLUMN: drop the column from stored columns from `final`.

At the end of the statement building, we run a post-processing step on the chain to eliminate redundant primary indexes. E.g. ALTER TABLE DROP COLUMN will only modify `final`, so `old`, `inter1`, `inter2` will have the same columns and thus we can remove `inter1` and `inter2`. The final primary index transition will then be `old` <-- `final`.

Fixes https://github.com/cockroachdb/cockroach/issues/99035
Epic: None
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/99526)
<!-- Reviewable:end -->

Xiang-Gu committed:
schemachanger: ADD COLUMN(s), DROP COLUMN(s), ALTER PK in one stmt

This commit supports ALTER TABLE stmt with any combination of
ADD COLUMN(s), DROP COLUMN(s), ALTER PK, and ADD CONSTRAINT commands in
the declarative schema changer.

The main idea is a two step process:
1). "inflate" a chain of four primary indexes where one is sourced from
its predecessor:
`old` <-- `inter1` <-- `inter2` <-- `final`
- Add column: add it to `inter1`, `inter2`, `final`
- Drop column: drop it from `final`
- Alter PK: alter pk in `inter2` and `final`

2). After processing all commands in this stmt, do a post-processing
to remove any redundant primary indexes:
- if `old` == `inter1`, remove `inter1`
- if `inter2` == `final`, remove `inter2`
- if `inter1` == `inter2`, remove the unremoved one or `final` if both
are removed

Release note (sql change): Users now (starting from v23.2) can issue one
ALTER TABLE statement with a combination of any number of ADD COLUMN,
any number of DROP COLUMN, one ALTER PRIMARY KEY, and any number of
ADD CONSTRAINT. For example, with this PR, we now support stmt like
```
CREATE TABLE t (i INT PRIMARY KEY, j INT NOT NULL, k INT NOT NULL);

ALTER TABLE t ADD COLUMN p INT DEFAULT 30, ALTER PRIMARY KEY USING COLUMNS (j), DROP COLUMN k, ADD CHECK (i > 0);
```

Xiang-Gu commented:
This should be ready for a look. I recognized that it is a big PR but I don't see an easy way to break it apart further. It is a major overhaul in the builder of several stmts (ADD COLUMN, DROP COLUMN, and ALTER PK). I think the easiest way to start reviewing is to understand the general approach as described in the PR first, and look for how it's exactly being implemented second in the code.

Thanks in advance.

postamar commented:
I like the goal of this PR very much. I don't think I'll have any objections of principle. I have a few comments before proceeding any further:
1. Can you spin off commits 1, 2 and 4 into a separate PR which can be merged quickly? Those changes are quite uncontroversial and I don't expect anything but nits 
2. Looking at commit 3, the one introducing `EXPLAIN (DDL, PK)`: I tried it out in a `cockroach demo` session after running the example in your release notes and all I'm seeing is a bunch of numbers: `Table 104: [1 : 2, 3]  <--  [1 : 2, 3, 4]  <--  [2 : 1, 3, 4]  <--  [2 : 1, 4]`. Can we have table and index and column names in there? It shouldn't be hard to generate placeholder names for nameless things. Also, the `treeprinter` package is quite easy to use, this output could be multiline. Finally, it's worth thinking about expanding this beyond just PKs; isn't what we really want here to know how many backfills (and perhaps also validations) the schema change is going to trigger? This seems like something that would be very useful to users whereas this PK flag as it is is only useful for the purposes of testing this PR. I don't want this to hold up this PR however. What do you think?
4. I'm finding the last commit a bit hard to review due to the large amount of noise in the diffs of the test output. That's the name of the game with these kinds of PRs but it seems that a large amount of noise is due to index and constraint IDs being bumped to higher values. In this optimization step, which removes redundant indexes, could we also remap the IDs of the new elements to make them contiguous with the existing ones? Like, if the NextIndexID index ID of a table is 3 before doing a vanilla ADD COLUMN (like in the `add_column` integration test), then the new primary index should have index ID 3 instead of 6. Regardless of the ease of review it's also the case that IDs jumping forward like this can be cause for confusion when trying to interpret an `EXPLAIN (DDL)` in production.
5. Following up on my previous question, would it be worth doing this optimization at pre-commit time instead? I'm very willing to believe that the answer to this question may in fact be "no" but it's something I was wondering about when implementing the generalized statement-time execution.

Xiang-Gu commented:
@postamar 
1. yes, those commits are now in this [PR](https://github.com/cockroachdb/cockroach/pull/103476).
2. Completely agreed! The initial purpose of me adding this additional EXPLAIN flag is to help me develop/debug this PR. I definitely liked your idea of extending it to include important stages (like BACKFILL and VALIDATE) with pretty table/index/column names. I think we can put all these into a separate, observability-improving PR.
3. 4. Let's chat tomorrow in our 1:1

Xiang-Gu commented:
@postamar Can I have some early review feedback for the latest push?

Per our discussion, the latest change has made an effort to keep the indexID same as before for "vanilla" ALTER TABLE cases (i.e. cases where we only add a column, or drop a column, or alter a PK and thus only one new primary index is created). 
The approach I used is to give "placeholder"/"temporary" IDs to newly added, possibly redundant primary indexes, and during post-processing, after we remove redundant primary indexes, we re-assign actual, sensible IDs to them. 

This has decreased some 100+ test file diffs and now it's some 300+ file diffs. I think we can leverage your `EXPLAIN(DDL, SHAPE)` and focus the test diff there. Those diffs are expected to me and I've noted some comments there too to help understand the diff. Hopefully, this effort helps you review this big PR.

Xiang-Gu commented on pkg/sql/schemachanger/testdata/explain_shape/add_column_default_unique:
The reason this file has changed is because:
1. We added the new primary index and gave it a "temporary"/"placeholder" indexID 
2. We added the new unique secondary index and gave it a real, actual indexID (2)
3. During post-processing, we replace the "placeholder" indexID with a real, actual indexID (4)

The diff is for temporary index so it's off-by-one.

Xiang-Gu commented on pkg/sql/schemachanger/testdata/explain_shape/alter_table_alter_primary_key_vanilla:
The diff here is because the code changed ALTER PK such that the new unique secondary index now uses final primary index as source to backfill (previously, it uses the old, current primary index as source, which, in general, in not correct, especially if we need more than one new primary index), so you see here backfilling the unique secondary index is delayer *after* we've backfilled the new primary index.

postamar commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_drop_column.go:
```diff
@@ -468,108 +470,180 @@ func panicIfColReferencedInPredicate(b BuildCtx, col *scpb.Column, tblElts Eleme
+// altered PK, so, (`old`, `inter1`, `inter2`, `final`).
+func getAllPrimaryIndexesSortedBySourcing(
+	b BuildCtx, tableID catid.DescID,
+) (old, inter1, inter2, final *scpb.PrimaryIndex) {
```
My impression reading a lot of the code in this package is that you might want some kind of struct instead of this tuple, and define methods on it.

postamar reviewed (commented):
I did a pass. Thanks for doing this! It's not easy. This seems like it should work but there's a few things that need clarifying before I review in detail. Thanks for doing the ID remappings, this PR is much easier to review now.

Xiang-Gu reviewed (commented):
Thanks @postamar. I've addressed most of your feedback except one (see my responses for detail). Appreciate it if you can give it a more detailed look and more feedback is welcome!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/99526#-:-NY5VfLx3Qc4jr37FZew:bu3pn1m)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball, @HonoreDB, and @postamar)

___
*[`pkg/sql/schemachanger/scbuild/builder_state.go` line 101 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/99526#gh-1231107781:-NXznpaB8Nv-P8aCmnmo:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f656b1238602051057ecaf4cd3a7a84b89622ea7/pkg/sql/schemachanger/scbuild/builder_state.go#L101)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

I don't think it's right to expose the user to this assertion failure, instead we should transparently fall back to the legacy schema changer so that they can get on with their lives. As far as we developers are concerned we certainly want to be informed about this, if we can have this message end up in the logs somehow we can look out for it in splunk.
</blockquote></details>

Done.
___
*[`pkg/sql/schemachanger/scbuild/builder_state.go` line 364 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/99526#gh-1231109077:-NXzzy5zC-UOgBo3GhzY:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f656b1238602051057ecaf4cd3a7a84b89622ea7/pkg/sql/schemachanger/scbuild/builder_state.go#L364)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Create a `nextTentativeIndexID` method instead. Boolean args are generally a code smell.
</blockquote></details>

Done.
___
*[`pkg/sql/schemachanger/scbuild/builder_state.go` line 374 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/99526#gh-1231109773:-NY--KSyAu9o5npk4Mye:bz4vy4v) ([raw file](https://github.com/cockroachdb/cockroach/blob/f656b1238602051057ecaf4cd3a7a84b89622ea7/pkg/sql/schemachanger/scbuild/builder_state.go#L374)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

`TABLE_TEMP_IDS_START` instead?
</blockquote></details>

done changed it to a const
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table.go` line 374 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/99526#gh-1231111681:-NY--bvG8MYGuEoC0EcV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f656b1238602051057ecaf4cd3a7a84b89622ea7/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table.go#L374)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Shouldn't this be a `const`?
</blockquote></details>

Done.
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table.go` line 324 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/99526#gh-1231146055:-NY4r74z5sgxkFEHLOWZ:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/b68fd1061020ff406aadd55013b917180430430a/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table.go#L324)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Define custom types instead of encoding meaning into strings. What you've done here isn't the worst thing in the world, of course, but leveraging the type system is a good habit to build.
</blockquote></details>

Done
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table.go` line 326 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/99526#gh-1231141518:-NY--amY1SKLf_HiHeB2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b68fd1061020ff406aadd55013b917180430430a/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table.go#L326)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

did you mean `nonNilPrimaryIndexes := make([]*scpb.PrimaryIndex, 0, 4)` ?
</blockquote></details>

Done.
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_add_column.go` line 259 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/99526#gh-1231119401:-NY4w3Mx4BCohGS060c8:bwkpb1x) ([raw file](https://github.com/cockroachdb/cockroach/blob/b68fd1061020ff406aadd55013b917180430430a/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_add_column.go#L259)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Why do we need to version-gate this? We're not adding any new elements or anything like that. If it's just because of the rules then we should backport the rules to 23.1 to ensure that in mixed-version clusters we can have a 23.1 node perform the schema change (and potentially roll it back) after the statement txn commits just like a 23.2 node would.
</blockquote></details>

I see your point of wanting to allow a 23.1 node pick up and complete a schema change job that was planned on 23.2 node.

yes, I think the reason I version gate it is because
1. I added three dep rules
2. I modified opgen file for temporary index 

I am not too familiar with the risk of backporting above two changes, and I'd definitely want to hear more of it from you. 
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_add_column.go` line 341 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/99526#gh-1231120690:-NY-C5oF2hD3Dz8y26QX:bnci1ja) ([raw file](https://github.com/cockroachdb/cockroach/blob/b68fd1061020ff406aadd55013b917180430430a/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_add_column.go#L341)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

What happens if we don't fall back here?
</blockquote></details>

With concurrent DML, this statement could end up rendering the table being un-readable. See https://github.com/cockroachdb/cockroach/issues/90174
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_primary_key.go` line 88 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/99526#gh-1231152970:-NY4r9GI7reIn4XjA5cl:b-m6abs0) ([raw file](https://github.com/cockroachdb/cockroach/blob/b68fd1061020ff406aadd55013b917180430430a/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_primary_key.go#L88)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Why disallow it? It's a dumb thing for a user to do, of course, but shouldn't the declarative schema changer be able to handle this gracefully without any extra code needed?
</blockquote></details>

If it's just the manipulation of columns in those newly added primary indexes, this is fine. What I haven't thought through is other changes that were made in the builder of ALTER PK. For example,
1. we recreate all secondary indexes in an ALTER PK because its keySuffix columns need to change to the new ones. If there are more than one ALTER PK command, then we need to revisit all of them and adjust their keySuffix columns;
2. we drop old shard column and/or old `rowid` column, and add possible new shard column if the new PK is requested to be sharded. If there are more than one ALTER PK command, then we also need to reconsider the modification of those elements.

So, really, what I thought is "ahh, no one should want to do this so let me try not to worry about those cases -- they don't seem trivial to implement so let me just throw an error and call it a day"
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_drop_column.go` line 583 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/99526#gh-1231156966:-NY-CSiA8PcU29rjeD1g:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b68fd1061020ff406aadd55013b917180430430a/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_drop_column.go#L583)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

should this be moved to helpers.go?
</blockquote></details>

Done.
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_drop_column.go` line 585 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/99526#gh-1231159273:-NY4jbOnALuTAmv8NWzP:b4p0u1d) ([raw file](https://github.com/cockroachdb/cockroach/blob/b68fd1061020ff406aadd55013b917180430430a/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_drop_column.go#L585)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

My impression reading a lot of the code in this package is that you might want some kind of struct instead of this tuple, and define methods on it.
</blockquote></details>

good point, I've defined a struct `primaryIndexChain` for this purpose with three important methods:
- inflate: inflate a chain by cloning one from its predecessor. This is called whenever we determine that we need to create at-least one new primary index to complete the schema change
- deflate: remove redundant new primary indexes, called at the end of each statement in the builder
- validate: called whenever we construct, inflate, and deflate a chain to assert it is a "valid" chain
___
*[`pkg/sql/schemachanger/scplan/internal/opgen/opgen_temporary_index.go` line 0 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/99526#gh-1231087371:-NY4uoOLD_hk6X__j4iX:bnvk7o1) ([raw file](https://github.com/cockroachdb/cockroach/blob/f656b1238602051057ecaf4cd3a7a84b89622ea7/pkg/sql/schemachanger/scplan/internal/opgen/opgen_temporary_index.go#L0)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

*(Reviewable was [unable](https://github.com/Reviewable/Reviewable/issues/14) to map this GitHub inline comment thread to the right spot — sorry!)*

It might be worth leaving a comment here. Normally, we mark any state transition that results in irretrievable information loss as non-revertible. In this case, going to DELETE_ONLY means we miss out on any concurrent inserts and updates. Presumably this is OK because the temp schema's state transitions are so tightly bound to the backfilled index?
</blockquote></details>

done, yes, it's because we have a dep rule that enforce a temporary index to not transition into DELETE_ONLY until its "master" index has transitioned into MERGED.
___
*[`pkg/sql/logictest/testdata/logic_test/alter_table` line 3327 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/99526#gh-1231093044:-NY-cCwi0rTWCLY9OZrU:b-zcjvte) ([raw file](https://github.com/cockroachdb/cockroach/blob/f656b1238602051057ecaf4cd3a7a84b89622ea7/pkg/sql/logictest/testdata/logic_test/alter_table#L3327)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Very nice.
</blockquote></details>

Thanks
___
*[`pkg/sql/opt/exec/execbuilder/testdata/not_visible_index` line 1240 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/99526#gh-1231092005:-NY4yUYV4-Gr_hRjHsvJ:bmb9a87) ([raw file](https://github.com/cockroachdb/cockroach/blob/f656b1238602051057ecaf4cd3a7a84b89622ea7/pkg/sql/opt/exec/execbuilder/testdata/not_visible_index#L1240)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Can we just do `SHOW INDEX FROM t1` from t1 here to keep to row order stable?
</blockquote></details>

I tried just `SHOW INDEX FROM t1` but it also gives a different order.
___
*[`pkg/sql/schemachanger/scbuild/testdata/alter_table_add_column` line 104 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/99526#gh-1231131361:-NY5BcwnCD4m2eYKU1le:bkcz002) ([raw file](https://github.com/cockroachdb/cockroach/blob/b68fd1061020ff406aadd55013b917180430430a/pkg/sql/schemachanger/scbuild/testdata/alter_table_add_column#L104)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

We shouldn't be creating a transient primary index here. It's not incorrect, just undesirable, and fundamentally avoidable. What are your thoughts on this?
</blockquote></details>

This is a really good find! I'd like to explain a bit more what happened and how I plan to address this.

So, the test is to call build on two add column statements, e.g.
```
create table t (i int primary key);

BUILD(alter table t add column j int default 30);
BUILD(alter table t add column k int default 40);
```

The first build:
1. inflate chain to `(i;) <-- (i;) <-- (i;) <-- (i;)`
2. add column j to make chain `(i;) <-- (i;j) <-- (i;j) <-- (i;j)`
2. deflate chain to `(i;) <-- (i;j)`

The second build comes along:
1. inflate the existing chain (incorrectly) to `(i;) <-- (i;) <-- (i;) <-- (i;j)`
2. add column k to make chain `(i;) <-- (i;k) <-- (i;k) <-- (i;j,k)`
3. deflate chain to `(i;) <-- (i;k) <-- (i;j,k)`

That's why we see an intermediate primary index that's being added. The culprit is the current inflate logic is too simplistic -- it simply iterate from old to inter1 to inter2 to final and clone one if it's nil from its predecessor. In this particular case, inflate for the second build should inflate the chain instead to `(i;) <-- (i;j) <-- (i;j) <-- (i;j)`. I've figured out the solution to change inflate logic and I will implement it once I come back next week on Tuesday.
___
*[`pkg/sql/schemachanger/scbuild/testdata/unimplemented_alter_table` line 0 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/99526#gh-1231088078:-NY5B3CX83IG6WzqQwoP:bw3497r) ([raw file](https://github.com/cockroachdb/cockroach/blob/f656b1238602051057ecaf4cd3a7a84b89622ea7/pkg/sql/schemachanger/scbuild/testdata/unimplemented_alter_table#L0)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

*(Reviewable was [unable](https://github.com/Reviewable/Reviewable/issues/14) to map this GitHub inline comment thread to the right spot — sorry!)*

Hooray :)

But now you'll need test cases for these in `pkg/sql/schemachanger/testdata`, perhaps. I'm not sure. Most important is that we have support in logic tests.
</blockquote></details>

The newly added logic test covers a lot more cases so unless you have a strong feeling here I will not add a end to end test here.
___
*[`pkg/sql/schemachanger/testdata/explain_shape/alter_table_alter_primary_key_vanilla` line 19 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/99526#gh-1231169932:-NY4kJgo64efjPK1FJvH:b1f4e3c) ([raw file](https://github.com/cockroachdb/cockroach/blob/b68fd1061020ff406aadd55013b917180430430a/pkg/sql/schemachanger/testdata/explain_shape/alter_table_alter_primary_key_vanilla#L19)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Why are we backfilling this index in a separate stage now? I looked at the EXPLAIN (DDL, VERBOSE) output and couldn't understand why.
</blockquote></details>

This is because we are now instructing this new unique secondary index to be backfilled from the new primary index, so it has to wait until the new primary index has been backfilled.


Previously, this new unique secondary index was backfilled from the old primary index. This is mostly okay except that the new primary index is on a newly added column, e.g. `alter table t add column j, alter primary key using columns (j)` in which case the old primary index does not have values of column `j` and thus cannot be used as source to backfill this new unique secondary index (i;j;)


<!-- Sent from Reviewable.io -->

postamar commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/dependencies.go:
```diff
@@ -226,15 +226,31 @@ type TableHelpers interface {
-	NextTableIndexID(table *scpb.Table) catid.IndexID
+	// `useTempID`, if true, returns a "temporary"/"placeholder" ID which is generated
+	// from a sequence starting from scbuildstmt.TableTentativeIdsStart. Otherwise, it returns
+	// the normal, actual ID that typically starts from 1.
```
nit: fix this comment now that you've gotten rid of `useTempID`

postamar commented on pkg/bench/rttanalysis/testdata/benchmark_expectations:
```diff
@@ -6,8 +6,8 @@ exp,benchmark
 15,AlterTableAddColumn/alter_table_add_1_column
-15,AlterTableAddColumn/alter_table_add_2_columns
-15,AlterTableAddColumn/alter_table_add_3_columns
+19,AlterTableAddColumn/alter_table_add_2_columns
```
Investigate this by adding a data-driven scbuild test. Here's what's going on:
```> EXPLAIN (DDL, SHAPE) ALTER TABLE alter_table ADD COLUMN a INT;                                                                
                                            info
--------------------------------------------------------------------------------------------
  Schema change plan for ALTER TABLE ‹defaultdb›.public.‹alter_table› ADD COLUMN ‹a› INT8;
   └── execute 3 system table mutations transactions
```
and
```
> EXPLAIN (DDL, SHAPE) ALTER TABLE alter_table ADD COLUMN a INT, ADD COLUMN b INT;                                              
                                                      info
-----------------------------------------------------------------------------------------------------------------
  Schema change plan for ALTER TABLE ‹defaultdb›.public.‹alter_table› ADD COLUMN ‹a› INT8, ADD COLUMN ‹b› INT8;
   ├── execute 2 system table mutations transactions
   ├── backfill using primary index alter_table_pkey- in relation alter_table
   │    └── into alter_table_pkey+ (rowid, a+, b+)
   ├── execute 2 system table mutations transactions
   ├── merge temporary indexes into backfilled indexes in relation alter_table
   │    └── from alter_table@[3] into alter_table_pkey+
   ├── execute 1 system table mutations transaction
   ├── validate UNIQUE constraint backed by index alter_table_pkey+ in relation alter_table
   └── execute 3 system table mutations transactions
```

postamar commented on pkg/sql/logictest/testdata/logic_test/alter_table:
```diff
@@ -3174,6 +3172,159 @@ table_name  constraint_name  constraint_type  details
+skipif config local-legacy-schema-changer
+skipif config local-mixed-22.2-23.1
+statement ok
+ALTER TABLE t_99035 ADD COLUMN k INT DEFAULT unique_rowid(), ALTER PRIMARY KEY USING COLUMNS (j), DROP COLUMN i, ADD COLUMN p INT DEFAULT 0, ADD CHECK (j>=0), ADD FOREIGN KEY (p) REFERENCES t_99035(j);
```
Add this as a schema changer end-to-end test case and make sure you exercise the DML injection test properly and I'll be happy.

postamar reviewed (approved):
LGTM, very nicely done. This is a major achievement which unlocks a lot of the schema changer's value.

This is nicely written and I'm reasonably confident this won't introduce too many regressions. 

I just have a couple of requests, other than that if CI is happy then I'm happy and you should go ahead and merge this.

Xiang-Gu committed:
sctest: quality improvement on mixed version related testing suites

This commit improves quality (readability mostly) of some of the
mixed version testing suites in declarative schame changer. Most
notably, it now skips test cases where stmt is not implemented but will
only be discovered in the builder.

Releaset note: None

Xiang-Gu commented:
TFTR!

CI is flaky so I am going to merge it...

bors r+
